src/java/boa/compiler/visitors/analysis/ReachingDefinitionGen.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2014, Hridesh Rajan, Robert Dyer, 
+ *                 and Iowa State University of Science and Technology
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package boa.compiler.visitors.analysis;
+
+import java.util.*;
+
+import boa.compiler.ast.Call;
+import boa.compiler.ast.Comparison;
+import boa.compiler.ast.Component;
+import boa.compiler.ast.Composite;
+import boa.compiler.ast.Conjunction;
+import boa.compiler.ast.Factor;
+import boa.compiler.ast.Identifier;
+import boa.compiler.ast.Index;
+import boa.compiler.ast.Node;
+import boa.compiler.ast.Operand;
+import boa.compiler.ast.Pair;
+import boa.compiler.ast.Selector;
+import boa.compiler.ast.Term;
+import boa.compiler.ast.UnaryFactor;
+import boa.compiler.ast.expressions.Expression;
+import boa.compiler.ast.expressions.SimpleExpr;
+import boa.compiler.ast.statements.AssignmentStatement;
+import boa.compiler.ast.statements.ReturnStatement;
+import boa.compiler.ast.statements.VarDeclStatement;
+import boa.compiler.visitors.*;
+
+/**
+ * @author rramu
+ */
+public class ReachingDefinitionGen extends AbstractVisitorNoArg {
+	ArrayList<String> variablesMonitored = new ArrayList<String>();
+	boolean inReturnStatement = false;
+	HashMap<Node,HashSet<String>> killVariable = new HashMap<Node,HashSet<String>>();
+	HashMap<Node,HashSet<String>> genVariable = new HashMap<Node,HashSet<String>>();
+
+	HashMap<Integer,Integer> gen = new HashMap<Integer,Integer>();
+	HashMap<String,HashSet<Integer>> defs = new HashMap<String,HashSet<Integer>>();
+
+	Node currentNode = null;
+	boolean killFlag = false, genFlag = false;
+
+	public final void dfs(final Node node, java.util.HashMap<Integer,String> nodeVisitStatus) {
```

Comment:
I'm a bit confused of the point of this (and the need for the nodeVisitStatus map).  The default visitor **does** a DFS visit of the tree.